File: questApplication/src/main/java/ru/nekit/android/qls/quest/QuestVisualBuilder.java
package ru.nekit.android.qls.quest;

import android.support.annotation.NonNull;
import android.view.View;

import ru.nekit.android.qls.quest.answer.CoinQuestAnswerChecker;
import ru.nekit.android.qls.quest.answer.GroupWeightComparisonQuestAnswerChecker;
import ru.nekit.android.qls.quest.answer.MetricsQuestAnswerChecker;
import ru.nekit.android.qls.quest.answer.SimpleExampleAnswerChecker;
import ru.nekit.android.qls.quest.answer.TrafficLightQuestAnswerChecker;
import ru.nekit.android.qls.quest.answer.common.QuestAnswerChecker;
import ru.nekit.android.qls.quest.common.Quest;
import ru.nekit.android.qls.quest.mediator.answer.ButtonsQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.content.EmptyQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.title.QuestTitleMediator;
import ru.nekit.android.qls.quest.mediator.types.choice.ChoiceAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.coins.CoinQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.coins.CoinQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.types.colored.ColoredVisualRepresentationQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.direction.DirectionQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.fruitArithmetic.FruitArithmeticQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.fruitArithmetic.FruitArithmeticQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.types.fruitArithmetic.FruitComparisonAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.metrics.MetricsQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.metrics.MetricsQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.types.perimeter.PerimeterQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.types.simpleExample.SimpleExampleQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.simpleExample.SimpleExampleQuestContentMediator;
import ru.nekit.android.qls.quest.mediator.types.textCamouflage.TextCamouflageContentMediator;
import ru.nekit.android.qls.quest.mediator.types.time.CurrentTimeQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.time.TimeQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.trafficLight.TrafficLightQuestAnswerMediator;
import ru.nekit.android.qls.quest.mediator.types.trafficLight.TrafficLightQuestContentMediator;

public class QuestVisualBuilder {

    @NonNull
    private final QuestContext mQuestContext;
    private IQuestMediatorFacade mQuestMediatorFacade;

    public QuestVisualBuilder(@NonNull QuestContext questContext) {
        mQuestContext = questContext;
    }

    public void create() {
        Quest quest = mQuestContext.generateQuest();
        switch (quest.getQuestType()) {

            case COINS:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new CoinQuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new CoinQuestContentMediator(),
                        new CoinQuestAnswerMediator()
                );

                break;

            case SIMPLE_EXAMPLE:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new SimpleExampleAnswerChecker(),
                        new QuestTitleMediator(),
                        new SimpleExampleQuestContentMediator(),
                        new SimpleExampleQuestAnswerMediator()
                );

                break;

            case METRICS:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new MetricsQuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new MetricsQuestContentMediator(),
                        new MetricsQuestAnswerMediator()
                );

                break;

            case PERIMETER:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new PerimeterQuestContentMediator(),
                        new SimpleExampleQuestAnswerMediator()
                );

                break;

            case TRAFFIC_LIGHT:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new TrafficLightQuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new TrafficLightQuestContentMediator(),
                        new TrafficLightQuestAnswerMediator()
                );

                break;

            case TEXT_CAMOUFLAGE:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new TextCamouflageContentMediator(),
                        new ButtonsQuestAnswerMediator()
                );

                break;

            case FRUIT_ARITHMETIC:

                boolean isSolution = quest.getQuestionType() == QuestionType.SOLUTION;
                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        isSolution ? new QuestAnswerChecker() :
                                new GroupWeightComparisonQuestAnswerChecker(),
                        new QuestTitleMediator(),
                        isSolution ? new FruitArithmeticQuestContentMediator() : null,
                        isSolution ? new FruitArithmeticQuestAnswerMediator() :
                                new FruitComparisonAnswerMediator()
                );

                break;

            case TIME:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        null,
                        new TimeQuestAnswerMediator()
                );

                break;

            case CURRENT_TIME:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        null,
                        new CurrentTimeQuestAnswerMediator()
                );

                break;

            case CHOICE:
            case MISMATCH:
            case CURRENT_SEASON:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        null,
                        new ChoiceAnswerMediator()
                );

                break;

            case COLORS:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        null,
                        new ColoredVisualRepresentationQuestAnswerMediator()
                );

                break;

            case DIRECTION:

                mQuestMediatorFacade = new QuestMediatorFacade(
                        mQuestContext,
                        new QuestAnswerChecker(),
                        new QuestTitleMediator(),
                        new EmptyQuestContentMediator(),
                        new DirectionQuestAnswerMediator()
                );

                break;

        }
        mQuestMediatorFacade.onCreate(mQuestContext);
    }

    public View getView() {
        return mQuestMediatorFacade.getView();
    }

    public void deactivate() {
        mQuestMediatorFacade.deactivate();
    }

    public void detachView() {
        mQuestMediatorFacade.detachView();
    }

}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/view/QuestVisualBuilder.kt
package ru.nekit.android.qls.quest.view

import io.reactivex.disposables.Disposable
import ru.nekit.android.qls.domain.useCases.QuestUseCases
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.quest.view.mediator.answer.ButtonListAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.content.EmptyQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.title.TitleMediator
import ru.nekit.android.qls.quest.view.mediator.types.choice.ChoiceAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.coins.CoinAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.coins.CoinQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.colored.ColoredVisualRepresentationAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.direction.DirectionAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.fruitArithmetic.FruitArithmeticAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.fruitArithmetic.FruitArithmeticQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.fruitArithmetic.FruitComparisonAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.metrics.MetricsAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.metrics.MetricsQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.perimeter.PerimeterQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.simpleExample.SimpleExampleAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.simpleExample.SimpleExampleQuestContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.textCamouflage.TextCamouflageContentMediator
import ru.nekit.android.qls.quest.view.mediator.types.time.CurrentTimeAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.time.TimeAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.trafficLight.TrafficLightAnswerMediator
import ru.nekit.android.qls.quest.view.mediator.types.trafficLight.TrafficLightQuestContentMediator
import ru.nekit.android.qls.shared.model.QuestType.*
import ru.nekit.android.qls.shared.model.QuestionType

//ver 1.2
class QuestVisualBuilder {

    companion object {

        fun build(questContext: QuestContext, body: (QuestMediatorFacade) -> Unit): Disposable =
                QuestUseCases.generateQuest().map { quest ->
                    val questMediatorFacade = when (quest.questType) {

                        COINS ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    CoinQuestContentMediator(),
                                    CoinAnswerMediator()
                            )

                        SIMPLE_EXAMPLE ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    SimpleExampleQuestContentMediator(),
                                    SimpleExampleAnswerMediator()
                            )

                        METRICS ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    MetricsQuestContentMediator(),
                                    MetricsAnswerMediator()
                            )

                        PERIMETER ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    PerimeterQuestContentMediator(),
                                    SimpleExampleAnswerMediator()
                            )

                        TRAFFIC_LIGHT ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    TrafficLightQuestContentMediator(),
                                    TrafficLightAnswerMediator()
                            )

                        TEXT_CAMOUFLAGE ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    TextCamouflageContentMediator(),
                                    ButtonListAnswerMediator()
                            )

                        FRUIT_ARITHMETIC -> {

                            val isSolution = quest.questionType == QuestionType.SOLUTION
                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    if (isSolution) FruitArithmeticQuestContentMediator() else
                                        null,
                                    if (isSolution)
                                        FruitArithmeticAnswerMediator()
                                    else
                                        FruitComparisonAnswerMediator()
                            )
                        }

                        TIME ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(), null,
                                    TimeAnswerMediator()
                            )

                        CURRENT_TIME ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(), null,
                                    CurrentTimeAnswerMediator()
                            )

                        CHOICE, MISMATCH, CURRENT_SEASON ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(), null,
                                    ChoiceAnswerMediator()
                            )

                        COLORS ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(), null,
                                    ColoredVisualRepresentationAnswerMediator()
                            )

                        DIRECTION ->

                            QuestMediatorFacade(
                                    questContext,
                                    TitleMediator(),
                                    EmptyQuestContentMediator(),
                                    DirectionAnswerMediator()
                            )
                    }
                    questMediatorFacade.onCreate(questContext, quest)
                    questMediatorFacade
                }.subscribe { it ->
                    body(it)
                }
    }
}